pkg/diff/diff.go

```diff
@@ -0,0 +1,691 @@
+	}
+
+	var oldDs datasourceTypes.DataSource
+	if err := json.Unmarshal([]byte(oldContent), &oldDs); err != nil {
+		return nil, errors.Wrapf(err, "json unmarshal")
+	}
+
+	var rawRepos = make(map[string]gitRepo)
+
+	for _, oldR := range oldDs.Raw {
+		name := strings.TrimSuffix(path.Base(oldR.URL), ".git")
+		g := gitRepo{
+			name:    name,
+			url:     oldR.URL,
+			dir:     filepath.Join(opts.dir, name),
+			old:     oldR.Commit,
+			remotes: make(map[string]remote),
+			filter:  opts.filter,
+			diffAlg: extracted.diffAlg,
+		}
+		log.Printf("[INFO] Working directory for raw: %s", g.dir)
+		if err := g.prepare(); err != nil {
+			return nil, errors.Wrapf(err, "init raw git repo. URL: %s", g.url)
+		}
+		rawRepos[name] = g
+	}
+
+	for _, newR := range newDs.Raw {
+		name := strings.TrimSuffix(path.Base(newR.URL), ".git")
+		if o, found := rawRepos[name]; found {
+			// Old data found, only overwrite "new" field
+			o.new = newR.Commit
+			rawRepos[name] = o
+			continue
+		}
+
+		// new side only
+		g := gitRepo{
+			name:    name,
+			url:     newR.URL,
+			dir:     filepath.Join(opts.dir, name),
+			new:     newR.Commit,
+			remotes: make(map[string]remote),
+			filter:  opts.filter,
+			diffAlg: extracted.diffAlg,
+		}
+		log.Printf("[INFO] Working directory for raw: %s", g.dir)
+		if err := g.prepare(); err != nil {
+			return nil, errors.Wrapf(err, "init raw git repo. URL: %s", g.url)
+		}
+		rawRepos[name] = g
+	}
+
+	return rawRepos, nil
+}
+
+func listRawfiles(extracted gitRepo, oldPath, newPath string) ([]rawfile, error) {
+	bs, err := extracted.showBlob(extracted.old, oldPath)
+	if err != nil {
+		return nil, errors.Wrapf(err, "git show. commit: %s, path: %s", extracted.old, oldPath)
+	}
+	var oldData dataTypes.Data
+	if err := json.Unmarshal([]byte(bs), &oldData); err != nil {
+		// Don't report error and just continue because older Data type does not include raw file paths.
+		// If dataTypes.Data struct will change, it'll be time to consider unmarhsalling it with interface{}.
+		log.Printf("[WARN] unmarshal of %s failed, probably data type is stale", oldPath)
+		return nil, nil
+	}
+
+	bs, err = extracted.showBlob(extracted.new, newPath)
+	if err != nil {
+		return nil, errors.Wrapf(err, "git show. commit: %s, path: %s", extracted.new, newPath)
+	}
+	var newData dataTypes.Data
+	if err := json.Unmarshal([]byte(bs), &newData); err != nil {
+		log.Printf("[WARN] unmarshal of %s failed, probably data type is stale", newPath)
+		return nil, nil
+	}
+
+	var rawfiles = make(map[string]rawfile)
+	for _, fullpath := range oldData.DataSource.Raws {
+		ss := strings.Split(fullpath, string(filepath.Separator))
+		if len(ss) == 0 {
+			return nil, errors.Errorf("raw path does not contain repository name. path: %s", fullpath)
+		}
+
+		rawfiles[fullpath] = rawfile{
+			repo: ss[0],
+			path: strings.Join(ss[1:], string(filepath.Separator)),
+			old:  true,
+		}
```

Comment:
This may not actually happen, but should we not take it into account when the length of ss is 1?

Also, if you want to cut at `/` and rejoin later, strings.Cut may be a better option.

Reply:
> This may not actually happen, but should we not take it into account when the length of ss is 1?

It works as https://go.dev/play/p/eDM0b6ZcZQL

> Also, if you want to cut at / and rejoin later, strings.Cut may be a better option.

Totally agree!

Reply:
strings.Cut version  6bea6f5